File: seeder lama/seed/user_seeder.py
from typing import List, Dict, Any
from sqlalchemy.orm import Session
from datetime import datetime

# Import dari project modules dengan try-except untuk robustness
try:
    from src.models.user import User, UserRole
    from src.config.hash import hash_password
    from src.database.seeder import BaseSeeder
except ImportError:
    # Fallback ke relative imports jika absolute gagal
    from ...models.user import User, UserRole
    from ...config.hash import hash_password
    from ..seeder import BaseSeeder

class UserSeeder(BaseSeeder):
    """Seed initial users into the database"""
    
    def __init__(self):
        super().__init__()
    def _generate_unique_id(self, user_id: int) -> str:
        """Generate unique_id untuk user"""
        timestamp = int(datetime.now().timestamp())
        # Combine user_id with timestamp and apply a simple transformation
        encoded_id = ((timestamp % 10000) * 100) + (user_id % 100)
        return f"USR-{encoded_id:06d}"
    
    def run(self):
        """Run user seeding"""
        try:
            print("🌱 Seeding users...")
            
            users_data = [
                {
                    "username": "admin",
                    "email": "admin@example.com",
                    "password": "AdminPass123!",
                    "nama" : "Site Admin",
                    "telpon": "123-456-7890",
                    "role": UserRole.ADMIN,
                    "is_active": True,
                    "is_verified": True,
                    "bio": "Administrator account"
                },
                {
                    "username": "moderator",
                    "email": "moderator@example.com",
                    "password": "ModPass123!",
                    "nama": "Site Moderator",
                    "telpon": "098-765-4321",
                    "role": UserRole.MODERATOR,
                    "is_active": True,
                    "is_verified": True,
                    "bio": "Moderator account"
                },
                {
                    "username": "johndoe",
                    "email": "john.doe@example.com",
                    "password": "Password123",
                    "nama": "John Doe",
                    "telpon": "123-456-7890",
                    "role": UserRole.USER,
                    "is_active": True,
                    "is_verified": False,
                    "bio": "Sample user account"
                },
                {
                    "username": "janedoe",
                    "email": "jane.doe@example.com",
                    "password": "Password123",
                    "nama": "Jane Doe",
                    "telpon": "321-654-0987",
                    "role": UserRole.USER,
                    "is_active": True,
                    "is_verified": True,
                    "bio": "Another sample user"
                }
            ]
            
            created_count = 0
            for i, user_data in enumerate(users_data, 1):
                existing_user = self.db.query(User).filter(
                    (User.email == user_data["email"]) | 
                    (User.username == user_data["username"])
                ).first()
                
                if not existing_user:
                    # Hash password before creating user
                    user_data["password"] = hash_password(user_data["password"])
                    
                    # Generate unique_id - Method 1: Predictable ID
                    user_data["unique_id"] = f"USR-{i:05d}"
                    
                    user = User(**user_data)
                    self.db.add(user)
                    self.db.flush()  # Flush untuk mendapatkan ID
                    
                    # Method 2: Update dengan ID asli (opsional)
                    # user.unique_id = f"USR-{user.id:05d}"
                    
                    created_count += 1
                    print(f"  ✅ Created user: {user_data['username']} ({user_data['email']}) - ID: {user_data['unique_id']}")
                else:
                    print(f"  ⚠️ User already exists: {user_data['username']} ({user_data['email']})")
            
            self.db.commit()
            print(f"✅ User seeding completed. Created {created_count} new users.")
            
        except Exception as e:
            self.db.rollback()
            raise Exception(f"User seeding failed: {e}")

# Utility functions tetap sama...
def create_user_if_not_exists(db: Session, data: Dict[str, Any]) -> User:
    """Helper function: Create a user if email/username not present."""
    existing = db.query(User).filter(
        (User.email == data["email"]) | (User.username == data["username"])
    ).first()
    
    if existing:
        return existing

    if not data["password"].startswith("$2b$"):
        data["password"] = hash_password(data["password"])

    # Generate unique_id jika tidak ada
    if "unique_id" not in data or not data["unique_id"]:
        # Get next available ID
        max_id = db.query(User.id).order_by(User.id.desc()).first()
        next_id = (max_id[0] + 1) if max_id else 1
        data["unique_id"] = f"USR-{next_id:05d}"

    user = User(**data)
    db.add(user)
    db.commit()
    db.refresh(user)
    return user

def seed_users(db: Session) -> List[User]:
    """Legacy function - use UserSeeder class instead"""
    seeder = UserSeeder()
    seeder.db = db
    seeder.run()
    return db.query(User).all()

File: cmd/seed/seeder/user_seeder.go
package seeder

import (
    "database/sql"
    "fmt"
    "github.com/jmoiron/sqlx"
    "REFACTORING_MAUNA/internal/domain/entities"
    "REFACTORING_MAUNA/pkg/security"
    "time"
)

type UserSeeder struct {
    db *sqlx.DB
}

func NewUserSeeder(db *sqlx.DB) *UserSeeder {
    return &UserSeeder{db: db}
}

func (s *UserSeeder) Name() string {
    return "UserSeeder"
}

type userData struct {
    Username   string
    Email      string
    Password   string
    Nama       string
    Telpon     string
    Role       entities.UserRole
    IsActive   bool
    IsVerified bool
    Bio        string
}

func (s *UserSeeder) Run() error {
    PrintInfo("Seeding users...")

    usersData := []userData{
        {
            Username:   "admin",
            Email:      "admin@example.com",
            Password:   "AdminPass123!",
            Nama:       "Site Admin",
            Telpon:     "123-456-7890",
            Role:       entities.RoleAdmin,
            IsActive:   true,
            IsVerified: true,
            Bio:        "Administrator account",
        },
        {
            Username:   "moderator",
            Email:      "moderator@example.com",
            Password:   "ModPass123!",
            Nama:       "Site Moderator",
            Telpon:     "098-765-4321",
            Role:       entities.RoleModerator,
            IsActive:   true,
            IsVerified: true,
            Bio:        "Moderator account",
        },
        {
            Username:   "johndoe",
            Email:      "john.doe@example.com",
            Password:   "Password123",
            Nama:       "John Doe",
            Telpon:     "123-456-7890",
            Role:       entities.RoleUser,
            IsActive:   true,
            IsVerified: false,
            Bio:        "Sample user account",
        },
        {
            Username:   "janedoe",
            Email:      "jane.doe@example.com",
            Password:   "Password123",
            Nama:       "Jane Doe",
            Telpon:     "321-654-0987",
            Role:       entities.RoleUser,
            IsActive:   true,
            IsVerified: true,
            Bio:        "Another sample user",
        },
    }

    // Start transaction
    tx, err := s.db.Beginx()
    if err != nil {
        PrintError(fmt.Sprintf("Failed to start transaction: %v", err))
        return err
    }
    defer func() {
        if err != nil {
            tx.Rollback()
        }
    }()

    createdCount := 0
    for i, user := range usersData {
        // Check if user already exists by email or username
        var existingID sql.NullInt64
        err := tx.QueryRow(
            "SELECT id FROM users WHERE email = $1 OR username = $2",
            user.Email,
            user.Username,
        ).Scan(&existingID)

        if err == nil && existingID.Valid {
            PrintWarning(fmt.Sprintf("User already exists: %s (%s)", user.Username, user.Email))
            continue
        }

        if err != nil && err != sql.ErrNoRows {
            PrintError(fmt.Sprintf("Database error: %v", err))
            return err
        }

        // Hash password
        hashedPassword, err := security.HashPassword(user.Password)
        if err != nil {
            PrintError(fmt.Sprintf("Failed to hash password: %v", err))
            return err
        }

        // Generate unique_id (format: USR-00001)
        uniqueID := fmt.Sprintf("USR-%05d", i+1)

        // Insert user
        _, err = tx.Exec(
            `INSERT INTO users (
                unique_id, username, email, password, nama, telpon, role, 
                is_active, is_verified, bio, created_at
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)`,
            uniqueID,
            user.Username,
            user.Email,
            hashedPassword,
            user.Nama,
            user.Telpon,
            user.Role,
            user.IsActive,
            user.IsVerified,
            user.Bio,
            time.Now(),
        )

        if err != nil {
            PrintError(fmt.Sprintf("Failed to create user: %v", err))
            return err
        }

        PrintSuccess(fmt.Sprintf("Created user: %s (%s) - ID: %s", user.Username, user.Email, uniqueID))
        createdCount++
    }

    // Commit transaction
    if err := tx.Commit(); err != nil {
        PrintError(fmt.Sprintf("Failed to commit transaction: %v", err))
        return err
    }

    PrintSuccess(fmt.Sprintf("User seeding completed. Created %d new users.", createdCount))
    return nil
}